src/Microsoft.AspNetCore.Rewrite/Internal/IISUrlRewrite/UrlRewriteFileParser.cs

```diff
@@ -127,70 +127,73 @@ private void ParseCondition(XElement condition, UrlRewriteRuleBuilder builder, P
-
-        private static void ThrowUrlFormatException(XElement element, string message)
-        {
-            var lineInfo = (IXmlLineInfo)element;
-            var line = lineInfo.LineNumber;
-            var col = lineInfo.LinePosition;
-            throw new FormatException(Resources.FormatError_UrlRewriteParseError(message, line, col));
-        }
-
-        private static void ThrowUrlFormatException(XElement element, string message, Exception ex)
-        {
-            var lineInfo = (IXmlLineInfo)element;
-            var line = lineInfo.LineNumber;
-            var col = lineInfo.LinePosition;
-            throw new FormatException(Resources.FormatError_UrlRewriteParseError(message, line, col), ex);
-        }
-
-        private static void ThrowParameterFormatException(XElement element, string message)
-        {
-            var lineInfo = (IXmlLineInfo)element;
-            var line = lineInfo.LineNumber;
-            var col = lineInfo.LinePosition;
-            throw new FormatException(Resources.FormatError_UrlRewriteParseError(message, line, col));
+            UrlAction action;
+            switch (actionType)
+            {
+                case ActionType.None:
+                    action = new NoneAction(stopProcessing ? RuleResult.SkipRemainingRules : RuleResult.ContinueRules);
+                    break;
+                case ActionType.Rewrite:
+                case ActionType.Redirect:
+                    var url = string.Empty;
+                    if (urlAction.Attribute(RewriteTags.Url) != null)
+                    {
+                        url = urlAction.Attribute(RewriteTags.Url).Value;
+                        if (string.IsNullOrEmpty(url))
+                        {
+                            throw new InvalidUrlRewriteFormatException(urlAction, "Url attribute cannot contain an empty string");
+                        }
+                    }
+
+                    var urlPattern = _inputParser.ParseInputString(url);
+                    var appendQuery = ParseBool(urlAction, RewriteTags.AppendQueryString, defaultValue: true);
+
+                    if (actionType == ActionType.Rewrite)
+                    {
+                        action = new RewriteAction(stopProcessing ? RuleResult.SkipRemainingRules : RuleResult.ContinueRules, urlPattern, appendQuery);
+                    }
+                    else
+                    {
+                        var redirectType = ParseEnum(urlAction, RewriteTags.RedirectType, RedirectType.Permanent);
+
+                        action = new RedirectAction((int)redirectType, urlPattern, appendQuery);
+                    }
+                    break;
+                case ActionType.AbortRequest:
+                    action = new AbortAction();
+                    break;
+                case ActionType.CustomResponse:
+                    int statusCode;
+                    if (!int.TryParse(urlAction.Attribute(RewriteTags.StatusCode)?.Value, out statusCode))
```

Comment:
hmm... ok. i thought about adding a range check but then wondered if that mattered since it was a custom response.